Feature/anonymous voting (#34)
vocdoni/docs: src/architecture/protocol/anonymous-voting/zk-census-proof.md

In our case:
  1. `Voter` is the owner of the `private key` corresponding to a certain `public key`
  2. `Voter`'s `public key` is included in the `census` Merkle Tree
  3. The nullifier provided by `Voter` uniquely corresponds to their `private key` and the `process ID` for a specific voting process

Although the computation is CPU and memory intensive, ZK Proofs can be generated from the user client running on modest hardware. The proof is validated by the Vochain Nodes, Miners, and any Third Party monitoring the process.

### zkSNARK circuit

Voters use the proving key generated for the circuit below to generate a ZK Proof. The circuit receives both private and public inputs.

Data that could reveal the identity of the voter are kept private (gray boxes in the diagram). Public inputs are submitted within the Vote Envelope, so that validators can check them against the proof and make sure that the user hasn't voted twice.

+ The same circuit can be used for any `process` with a census size of the same range (10k, 100k, 1M, etc).

> + It relies on a **trusted setup ceremony**

Reviewer: Could you link to somewhere to read about what a trusted setup is?

Author: Right! Added a link to an article

7. *[V]* StartBlock is reached, **election starts**
8. *[V]* Last merkle tree root becomes **censusRoot**
9. *[V+U]* **Get MerkleProof**
    - Vochain will send the *'compressed MerkleProof'* (which are the siblings compressed)
    - Client side will need to 'decompress' it
        - The logic to decompress the siblings [can be found here](https://github.com/vocdoni/arbo/blob/master/tree.go#L606)
        - And [here](https://github.com/vocdoni/arbo/blob/master/tree.go#L577) the explaination of the encoding
10. *[U+G]* **Get ProvingKey & WitnessCalc**
    - *Proving Key & Witness Calc* depend on the circuit being used
11. *[U]* **Generate zkInputs**
    - check the [zkInputs generation](#zkInputs-generation) section for more details
12. *[U]* **Generate zkSNARK proof**
    - using: *zkInputs + Proving Key + Witness Calculator*
13. *[U]* **Cast the vote** with zkSNARK proof
    - contains:
        - *public inputs*
        - *zkProof*
        - *vote*
14. *[V]* **Verify zkSNARK proof**, accept the vote
    - Needs to know:
        - ElectionID
        - Verification Key (depends on the circuit being used (census size))
        - User's *public inputs* + *zkProof*

### Merkle Tree

> The MerkleTree needs to be a zkSNARK-friendly implementation. As currently we are using [Circom]() for the zkSNARK circuits, we need to be compatible with the [circomlib](https://github.com/iden3/circomlib/tree/master/circuits/smt) MerkleTree implementation. A specification of the MerkleTree [can be found here](https://docs.iden3.io/publications/pdfs/Merkle-Tree.pdf).

Reviewer: `[Circom]()` needs a link destination

Author: Right! Updated

The following diagram contains a visual representation of the data structure of the Leaves of the MerkleTree being used in the scheme of the zk-census-proof.

![](./zk-census-proof-poseidon-merkletree-diagram.png)

### zkInputs generation
```json=
// Example of zkInputs
{
	"censusRoot": "51642541620950251760298704744678482162425252475654827255045491135352807540162",
	"censusSiblings": ["0","0","0","0"],
	"privateKey": "6190793965647866647574058687473278714480561351424348391693421151024369116465",
	"voteValue": "1",
	"electionId": "10",
	"nullifier": "1938187656076799017313903315498318464349291455761501098436114043715056719301",
}
```

Origin of each zkInput parameter:
> *all the parameters are `string` or `[]string` that represent `bigInt` or `[]bigInt`*
- *censusRoot*: computed by the *CensusAuthority* from the *Census Tree*
- *censusSiblings*: computed by the *CensusAuthority*, is the *Merkle Proof*
    - the *User* retrieves the *siblings* from the *Vochain* through the *Gateway*
    - the length of *censusSiblings* will depend on the *zkCircuit*:

>         - The design of the *MerkleTree* used in circomlib provokes different lengths in the siblings returned when generating a *MerkleProof*

Reviewer: I'm confused what this line means.

Author: Added a bit of more explanation and a link to the spec of the merkletree

Circuits are identified across the stack by using a Protobuf enum type. Each `ProofZkSNARK` protobuf package will have a `Type` identifier indicating which circuit the proof belongs to, so the Vochain knows which `Verification Key` to use for verifying the proof.

Format: `CIRCUITNAME_PARAMETER1_PARAMETER2`

List of current types:
- `ZKCENSUSPROOF_NLEVELS`
    - Circuit name: [`ZKCENSUSPROOF`](https://github.com/vocdoni/zk-franchise-proof-circuit/blob/master/circuits/census.circom)
    - Parameters: `nLevels`
    - Example:
        - `ZKCENSUSPROOF_100`
        - `ZKCENSUSPROOF_1000`

# Annex
#### KeyKeepers reveal and commit keys

A set of commitment keys are generated for each election process by a set of trusted identities named `keykeepers`. Only if all `keykeepers` are malicious could they tamper with the election, so it is crucial to distribute these special identities well. Once all these keys are revealed, anyone can generate a valid proof. This mechanism is added to the circuit in order to avoid vote buying when the election is over. Since anyone can now generate a valid proof, a voter will no longer be able to prove that they are the owner of a specific vote nullifier.

> ![](./circuit-with-keykeepers-keys-diagram.png)

Reviewer: what software did you use to generate this diagram? Could it be fetched directly from the source so that the designs are also editable by anyone? 
For example, many of the diagrams are created with draw.io and hosted on the https://github.com/vocdoni/design/tree/main/docs repo so they can be edited without re-downloading

Author: Yes, it's done with draw.io, and when saving the image marked the checkbox to include the editable data inside the png for future editions.
Didn't knew about a separated repository for the diagrams. Maybe these kind of diagrams are dependent on what the `docs` repo has as content and would be more useful with the images in the same repo & dir than the `.md` file? (including the images that are editable from drawio, like this one)
For example, this current image, it can live here in this directory, and if there is one change in the design of the protocol we just need to add 1 commit in this repo updating the `md` files and the drawio image. While if we need to update the `md` of this repo and then the image of the `design` repo, would be 2 commits in two different repos, with the dependency that for example the PR in the `design` repo with the new image passes, but in the `docs` repo maybe somebody does a comment with a suggestion of change of the spec, so the `docs` PR needs to be updated, but the `design` repo PR is already merged. This could be avoided by adding some kind of 'human protocol' where if somebody creates a new PR that depends on another PR of another repo there should be some kind of coordination, but in the case of these images, maybe it's easier & faster to just have the images in the same directory where it is the document which contains the image.
But maybe I'm missing some reason behind to have the images in a separated repo, so I can move this image into that repo if needed.

Reviewer: yes, you're right. I'll work on moving all the designs into this repo. Is the source for this design in this repo?

Author: Yes, the diagram source is embedded in the png image itself.

# ZK Census Proof

> This document is splitted in two main sections:

Reviewer 2: ```suggestion
This document is divided in two sections:
```

> - **Protocol design**: high level overview of the protocol, without giving much specific details, but showing the main idea.

Reviewer 2: ```suggestion
- **Protocol design**: high level overview of the protocol.
```

> - **Implementation spec**: step by step specification, explaining all the details, the data structures used, how things are hashed, etc. With the aim to avoid leaving any ambiguous concept.

Reviewer 2: ```suggestion
- **Implementation spec**: detailed specification, featuring the relevant data structures, how things are hashed, etc.
```

## Protocol design

> The census proof (also called franchise proof) enables user privacy and allows anonymous voting.

Reviewer 2: The `census proof` also applies to signed votes, doesn't it?

> The starting point is a [Merkle Proof](/architecture/census-overview?id=the-census), which efficiently proves that a public key belongs to a Merkle Tree (census). However, using this proof alone would allow the organizer to correlate the vote envelopes with the public key on the database, so that votes wouldn't be secret.

Reviewer 2: I think the link is now `architecture/census-overview.html`

Author: Updated

To this end, Vocdoni achieves voting anonymity by the use of ZK-Snarks.

### zkSNARKs for anonymous voters

Snark stands for *Succinct non-interactive argument of knowledge*. In our case, this means proving to someone that we know something, but without revealing it.

In our case:
  1. `Voter` is the owner of the `private key` corresponding to a certain `public key`
  2. `Voter`'s `public key` is included in the `census` Merkle Tree
  3. The nullifier provided by `Voter` uniquely corresponds to his/her `private key` and the `process ID`

Although the computation is CPU and memory intensive, ZK Proofs can be generated from the user client running on modest hardware. The proof is validated by the Vochain Nodes, Miners and any Third Party monitoring the process.

### zkSNARK circuit

Voters use the proving key generated for the circuit below to generate a ZK Proof. The circuit receives both private and public inputs.

Data that could reveal the identity of the voter are kept private (gray boxes in the diagram). Public inputs are submitted within the Vote Envelope, so that validators can check them against the proof and make sure that the user hasn't voted twice.

> + The same circuit can be used for any `process` with the same census (max) size

Reviewer 2: ```suggestion
+ The same circuit can be used for any `process` with a census size of the same range (10k, 100k, 1M, etc).
```

Vochain->>Vochain: 7. startBlock is reached, election starts
Vochain->>Vochain: 8. last merkle tree root becomes censusRoot
Vochain->>User: 9.get MerkleProof
# note, next line goes from User->User instead of User->IPFS to avoid spending a full column just for 1 query to IPFS
User->>User: 10. get ProvingKey & WitnessCalc (from IPFS GW)
User->>User: 11. generate zkInputs
User->>User: 12. generate zkSNARK proof
User->>Vochain: 13. cast the vote with zkSNARK proof
Vochain->>Vochain: 14. verify zkSNARK proof, accept the vote
```

#### Steps description
> - *[O]* = Organizer
> - *[U]* = User
> - *[V]* = Vochain
> - *[G]* = Gateway

0. Circom [circuit](https://github.com/vocdoni/zk-franchise-proof-circuit) is compiled & **Trusted Setup** generated
1. *[O]* Create user **login keys**
    - from csv data (+ user_secret?)
2. *[O]* Build **merkleTree** with login keys
3. *[O+V]* Create **new voting process** (newProcessTx)
4. *[U]* Generate **voting key** (used as leaf key)
    - User's *private key*: `sk = Hash(login_key | user_secret)`
    - User's *public key*: `pk = eddsa.prv2pub(sk)`
        - method: [`eddsa.prv2pub(sk)`](https://github.com/iden3/circomlib/blob/master/src/eddsa.js#L34)

>     - User's *voting key* (censustree leaf key): `leafKey = poseidon(pk.x, pk.y)`

Reviewer 2: ```suggestion
    - User's *voting key* (censustree leaf key): `leafKey = poseidon([pk.x, pk.y])`
```

        - This is the key that will be added into the *CensusTree*
5. *[U+V]* **Register voting key** using **login key** (registerKeyTx)
6. *[V]* Build voting **census merkle tree** (in state)

>     - Where each leaf contains each hash of user's `pk` (*voting key*)

Reviewer 2: ```suggestion
    - Where each leaf contains the hash of each user's `pk` (*voting key*)
```

    - contains:
        - *public inputs*
        - *zkProof*
        - *vote*
14. *[V]* **Verify zkSNARK proof**, accept the vote
    - Needs to know:
        - ElectionID
        - Verification Key (depends on the circuit being used (census size))
        - User's *public inputs* + *zkProof*

### Merkle Tree
The MerkleTree needs to be a zkSNARK friendly implementation. As currently we are using [Circom]() for the zkSNARK circuits, we need to be compatible with the [circomlib](https://github.com/iden3/circomlib/tree/master/circuits/smt) MerkleTree implementation. A specification of the MerkleTree [can be found here](https://docs.iden3.io/publications/pdfs/Merkle-Tree.pdf).

In the Vochain, we're using the [arbo](https://github.com/vocdoni/arbo) MerkleTree, which is a Go implementation compatible with the Circom spec.

The MerkleTree uses the Poseidon hash, which is a 'snark-friendly' hash function that later on can be proved inside a circuit without requiring too many constraints.

In the following diagram there is a visual representation of the data structure of the Leafs of the MerkleTree being used in the scheme of the zk-census-proof.

> ![](https://i.imgur.com/RVnFDqP.png)

Reviewer 2: Wouldn't it make sense to upload this file here?

Author: Agree, for the moment was a hackmd generated link as this content was copy&pasted here directly from hackmd each time that the hackmd was updated.
Added the original image (which contains the draw.io editable data source) & updated the link to it.

### zkInputs generation
```json=
// Example of zkInputs
{
	"censusRoot": "51642541620950251760298704744678482162425252475654827255045491135352807540162",
	"censusSiblings": ["0","0","0","0"],
	"privateKey": "6190793965647866647574058687473278714480561351424348391693421151024369116465",
	"voteSigS": "2093461910575977345603199789919760192811763972089699387324401771367839603655",
	"voteSigR8x": "8553678144208642175027223770335048072652078621216414881653012537434846327449",
	"voteSigR8y": "5507837342589329113352496188906367161790372084365285966741761856353367255709",
	"voteValue": "1",
	"electionId": "10",
	"nullifier": "1938187656076799017313903315498318464349291455761501098436114043715056719301",
	"relayerPublicKey": "100",
	"relayerProof": "21349690342514405503176665977362532634490340702670001813783738965751319356478",
	"revealKey": ["0"],
	"commitKey": ["19014214495641488759237505126948346942972912379615652741039992445865937985820"]
}
```

Origin of each zkInput parameter:
> *all the parameters are `string` or `[]string` that represent `bigInt` or `[]bigInt`*

> - *censusRoot*: computed by the *CensusAuthorithy* from the *Census Tree*

Reviewer 2: Who is the `CensusAuthorithy`? Maybe we can explain this a bit better? (two step process, this is the second Merkle Tree, etc) ?

> - *censusSiblings*: computed by the *CensusAuthorithy*, it's the *Merkle Proof*

Reviewer 2: Comment that the proof is retrieved from the Vochain? What kind of Merkle Tree is it (arbo)? etc

Author: Added a line regarding the Vochain origin of the siblings data. The MerkleTree type is defined in the `Merkle Tree` section

        - The logic needed to be implemented in the User side can be found [here (go) lines 67-70](https://github.com/vocdoni/zk-franchise-proof-circuit/blob/feature/go-code-inputs-generation/test/go-inputs-generator/census_test.go#L67), and [here (js) line 23](https://github.com/vocdoni/zk-franchise-proof-circuit/blob/feature/go-code-inputs-generation/src/franchise.js#L33):
            - `while (siblings.length < this.levels) siblings.push(BigInt(0));`
- *privateKey*: generated by the *User*
    - code sample: [franchise.js#L48](https://github.com/vocdoni/zk-franchise-proof-circuit/blob/master/src/franchise.js#L48):
        ```js
        const rawpvkHash = eddsa.pruneBuffer(createBlakeHash("blake512").update(rawpvk).digest().slice(0, 32));
        const privateKey = Scalar.shr(ffutils.leBuff2int(rawpvkHash),3);
        ```
- *voteSigS*: parameter *S* from the signature computed by the *User*
    - `voteSig = sk.SignPoseidon(vote)`
- *voteSigR8x*: coordinate *X* from the signature point *S* (computed by the *User*)
- *voteSigR8y*: coordinate *Y* from the signature point *S* (computed by the *User*)

> - *voteValue*: value of the *User* vote

Reviewer 2: How do we serialize/compute it? The origin is a JSON payload

Author: This would need to be discussed. For the moment in the circuit related code, it's assumed that the `voteValue` is a biginteger.

- *electionId*: the election ID in which the *User* is participating
- *nullifier*: computed by *User*
    - `nullifier = poseidon.Hash(sk, electionID)`
- *relayerPublicKey*: given by the *KeyKeeper*
- *relayerProof*: computed by the *User*

>     - `relayerProof = poseidon.Hash(nullifier, relayerPublicKey)`

Reviewer 2: Are we considering ZK rollups and relayers at this point?

> - *revealKey*: not known by the user at the proof generation moment

Reviewer 2: Is there a particular order in which they should appear? When there is no reveal key published, should the input still be `["0"]` and not `["0", "0", ... ]` ?

Author: The length of the arrays `relayerProof` & `commitKey` are determined by the circuit parameter `nMiners`.
(Adding this description to the doc)

- *commitKey*: given by the *KeyKeeper*
    - `poseidon.Hash(relayerPublicKey)`

### Circuit identification
There will be different circuits of the `zk-census-proof` depending on the census size, also there might be more use cases with different circuit designs.

> The client & Vochain need a way to unequivocally identify those circuits, in order to user the proper `Proving key`, `Witness calculator` and `Verification key` for each circuit.

Reviewer 2: ```suggestion
Both the client and the Vochain need a way to univocally identify those circuits, in order to user the proper `Proving key`, `Witness calculator` and `Verification key` for each circuit.
```

> The following defines the way how the circuits are identified in the client and also inside the Vochain through a Protobuf enum type. Each `ProofZkSNARK` protobuf packet will have a `Type` identifier that states to which circuit belongs the proof, so the Vochain knows which `Verification Key` to use to verify the proof.

Reviewer 2: ```suggestion
Circuits are identified across the stack by using a Protobuf enum type. Each `ProofZkSNARK` protobuf package will have a `Type` identifier indicating which circuit the proof belongs to, so the Vochain knows which `Verification Key` to use for verifying the proof.
```
